v0/destinations/marketo/transform.js

```diff
@@ -293,16 +288,110 @@ const processEvent = async (message, destination) => {
 };
 
 const process = async event => {
-  const response = await processEvent(event.message, event.destination);
+  const token = await getAuthToken(formatConfig(event.destination));
+  if (!token) {
+    throw Error("Authorisation failed");
+  }
+  const response = await processEvent(event.message, event.destination, token);
   return response;
 };
 
-const processRouterDest = input => {
-  const inputs = input;
-  inputs.batched = false;
-  inputs.statusCode = 200;
-  inputs.metadata = [input.metadata];
-  return inputs;
+/**  create output according to  {
+      "message": {},
+      "metadata": [],
+      "batched": false,
+      "statusCode": 500,
+      "error": "",
+      "destination": { "ID": "a", "url": "a" }
+    }
+    */
+// Success responses
+const getSuccessRespEvents = (message, metadata, destination) => {
+  const returnResponse = {};
+  returnResponse.message = message;
+  returnResponse.metadata = metadata;
+  returnResponse.batched = false;
+  returnResponse.statusCode = 200;
+  returnResponse.destination = destination;
+  return returnResponse;
+};
+
+// Error responses
+const getErrorRespEvents = (metadata, statusCode, error) => {
+  const returnResponse = {};
+  returnResponse.metadata = metadata;
+  returnResponse.batched = false;
+  returnResponse.statusCode = statusCode;
+  returnResponse.error = error;
+  return returnResponse;
+};
+
+const processRouterDest = async input => {
+  // Token needs to be generated for marketo which will be done on input level.
+  // If destination information is not present Error should be thrown
+  if (!Array.isArray(input) || input.length <= 0) {
+    const respEvents = getErrorRespEvents(
+      null,
+      400,
+      "Destination config not present for event"
+    );
+    return [respEvents];
+  }
+  let token;
+  try {
+    token = await getAuthToken(formatConfig(input[0].destination));
+  } catch (error) {
+    const respEvents = getErrorRespEvents(
+      input.map(ev => ev.metadata),
+      error.response ? error.response.status : 400,
+      error.message || "Error occurred while processing payload."
+    );
+    return [respEvents];
+  }
+  // If token is null track/identify calls cannot be executed.
+  if (!token) {
+    const respEvents = getErrorRespEvents(
+      input.map(ev => ev.metadata),
+      400,
+      "Authorisation failed"
+    );
+    return [respEvents];
+  }
+  const respList = [];
+  // Checking previous status Code. Initially setting to false.
+  // If true then previous status is 500 and every subsequent event output should be sent with status code 500 to the router to be retried.
+  let prevStatus = false;
+  for (let i = 0; i < input.length; i += 1) {
+    const inputs = input[i];
+    let respEvents = {};
+    if (prevStatus) {
+      respEvents = getErrorRespEvents(
+        [inputs.metadata],
+        500,
```

Comment:
this case will never hit right?
since we are making an external call at the very first, do we need this?

Reply:
if we need, can we do a continue instead of if-else